### ISSUE TITLE
priority of `do_sample` and `temperature` parameters in text generation

### ISSUE THREAD

**Issue by stanleee5** [huggingface/text-generation-inference#1201]
text generation client has `temperature` to module the logits distribution, and has `do_sample` option to Activate logits sampling.

i thought do_sample option has higher priority to choose decoding method (greedy/sampling). 
However in code, it changes by temperature parameters.
(if i give both do_sample=False and temperature value (e.g 0.2), texts are generated in sampling mode)

https://github.com/huggingface/text-generation-inference/blob/v1.1.0/server/text_generation_server/utils/tokens.py#L187

I wonder if this part was intentional.

**Comment by OlivierDehaene** [comment]
Yes this is intentional.
Modifying the temparture has no effect on greedy so if the user set it we default to do_sample true.
We could add a validation check to make sure that do_sample != false though.

**Comment by stanleee5** [comment]
Thank you for answering! 
I think validation check should be placed at Rust which is not familiar to me, I would check the code first.
Since the code/logic is correct, I will close the issue

**Comment by v-prgmr** [comment]
In order to force deterministic output generation, what needs to be done? @stanleee5

**Comment by stanleee5** [comment]
@v-prgmr I am using do_sample=False, temperature=None, and top_p=None for huggingface_hub InferenceClient's text_generation() parameter

**Comment by zagyvad** [comment]
@OlivierDehaene I noticed that current behaviour is: 
- do_sample=False, temperature=None -> greedy decoding 
- do_sample=False, temperature=0.9 (or anything other than 1) -> multinomial sampling
- do_sample=True, temperature=1 -> multinomial sampling
- do_sample=False, temperature=1 -> greedy decoding (with the logic you described above this should be also multinomial sampling)

I would recommend to change this and if do_sample=False to ignore the values for sampling related parameters like temperature. This way it would be consistent with transformers GenerationConfig

**Comment by mapa17** [comment]
Hi, I fail to recreate the behavior posted by @zagyvad for TGI 1.2
The only way to get deterministic output is setting a `seed` value.

Is the relevant code for this discussion the following? https://github.com/huggingface/text-generation-inference/blob/3238c49121b02432bf2938c6ebfd44f06c5adc2f/server/text_generation_server/utils/tokens.py#L186C1-L211C35

If so, do_sample is overwritten by any value in temperature, top_k, top_p, typical_p?

For me to achieve deterministic behavior I have to set
- do_sample=False, temperature=None, top_k=None, top_p=None, typical_p=None

With using temperature values
- (0, 1.0) Sharpening the distribution
- 1.0 - No temperature
- (1.0, inf) Becoming uniform with higher values

Can someone confirm or correct this please?
Thank you
